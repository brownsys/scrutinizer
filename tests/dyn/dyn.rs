#![allow(dead_code)]

mod object_type_eraser {
    trait DynamicTrait {
        fn inc(&self, a: usize) -> usize;
    }

    struct PureIncrementer;

    struct ImpureIncrementer;

    impl DynamicTrait for PureIncrementer {
        fn inc(&self, a: usize) -> usize {
            a + 1
        }
    }

    impl DynamicTrait for ImpureIncrementer {
        fn inc(&self, a: usize) -> usize {
            println!("{}", a);
            a + 1
        }
    }

    #[doc = "impure"]
    fn dyn_eraser(a: usize) -> usize {
        let dynamic: &dyn DynamicTrait = if a == 0 {
            &PureIncrementer {}
        } else {
            &ImpureIncrementer {}
        };
        dyn_eraser_helper(a, dynamic)
    }

    #[doc = "impure"]
    fn dyn_eraser_helper(a: usize, dynamic: &dyn DynamicTrait) -> usize {
        dynamic.inc(a)
    }
}

mod returns_impl_fn {
    #[doc = "pure"]
    fn outer(a: usize) -> usize {
        let cl = hof(a);
        execute(a, &cl)
    }

    #[doc = "impure"]
    fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
        cl(a)
    }

    #[doc = "pure"]
    fn hof(a: usize) -> impl Fn(usize) -> usize {
        move |x| x + a
    }
}

mod passthrough_impl_fn {
    #[doc = "pure"]
    fn outer(a: usize) -> usize {
        let cl = hof(a);
        execute(a, identity(&cl))
    }

    #[doc = "impure"]
    fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
        cl(a)
    }

    #[doc = "pure"]
    fn hof(a: usize) -> impl Fn(usize) -> usize {
        move |x| x + a
    }

    #[doc = "pure"]
    fn identity<T>(a: T) -> T {
        a
    }
}

mod returns_boxed_fn {
    #[doc = "pure"]
    fn outer(a: usize) -> usize {
        let cl = hof(a);
        execute(a, &cl)
    }

    #[doc = "impure"]
    fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
        cl(a)
    }

    #[doc = "pure"]
    fn hof(a: usize) -> Box<dyn Fn(usize) -> usize> {
        Box::new(move |x| x + a)
    }
}

mod returns_impl_fn_with_upvars {
    #[doc = "pure"]
    fn outer(a: usize) -> usize {
        let lam = |x| x + 1;
        let cl = hof(a, &lam);
        execute(a, &cl)
    }

    #[doc = "impure"]
    fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
        cl(a)
    }

    #[doc = "impure"]
    fn hof(a: usize, cl: &dyn Fn(usize) -> usize) -> impl Fn(usize) -> usize + '_ {
        move |x| cl(x + a)
    }
}

mod returns_boxed_fn_with_upvars {
    #[doc = "pure"]
    fn outer(a: usize) -> usize {
        let lam = |x| x + 1;
        let cl = hof(a, &lam);
        execute(a, &cl)
    }

    #[doc = "impure"]
    fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
        cl(a)
    }

    #[doc = "impure"]
    fn hof(a: usize, cl: &dyn Fn(usize) -> usize) -> Box<dyn Fn(usize) -> usize + '_> {
        Box::new(move |x| cl(x + a))
    }
}

fn main() {}
