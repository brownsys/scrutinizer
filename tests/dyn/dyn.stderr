error: static analysis was not able to verify the purity of the region
  --> dyn.rs:26:5
   |
26 | /     fn dyn_eraser(a: usize) -> usize {
27 | |         let dynamic: &dyn DynamicTrait = if a == 0 {
28 | |             &PureIncrementer {}
29 | |         } else {
...  |
32 | |         dyn_eraser_helper(a, dynamic)
33 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region
   = note: `#[deny(scrutinizer_lint)]` on by default

error: static analysis was not able to verify the purity of the region
  --> dyn.rs:36:5
   |
36 | /     fn dyn_eraser_helper(a: usize, dynamic: &dyn DynamicTrait) -> usize {
37 | |         dynamic.inc(a)
38 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
  --> dyn.rs:49:5
   |
49 | /     fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
50 | |         cl(a)
51 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
  --> dyn.rs:67:5
   |
67 | /     fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
68 | |         cl(a)
69 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
  --> dyn.rs:90:5
   |
90 | /     fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
91 | |         cl(a)
92 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
   --> dyn.rs:109:5
    |
109 | /     fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
110 | |         cl(a)
111 | |     }
    | |_____^
    |
    = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
   --> dyn.rs:114:5
    |
114 | /     fn hof(a: usize, cl: &dyn Fn(usize) -> usize) -> impl Fn(usize) -> usize + '_ {
115 | |         move |x| cl(x + a)
116 | |     }
    | |_____^
    |
    = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
   --> dyn.rs:128:5
    |
128 | /     fn execute(a: usize, cl: &dyn Fn(usize) -> usize) -> usize {
129 | |         cl(a)
130 | |     }
    | |_____^
    |
    = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
   --> dyn.rs:133:5
    |
133 | /     fn hof(a: usize, cl: &dyn Fn(usize) -> usize) -> Box<dyn Fn(usize) -> usize + '_> {
134 | |         Box::new(move |x| cl(x + a))
135 | |     }
    | |_____^
    |
    = help: consider using sandbox or privacy region
